crates/projection/src/lib.rs

```diff
@@ -70,20 +71,24 @@ impl Projection for CameraMatrix {
         z: f32,
     ) -> Result<Point2<f32>, Error> {
         let camera_ray = self.pixel_to_camera(pixel_coordinates);
-        let camera_ray_over_ground = self.camera_to_ground.rotation * camera_ray;
-        if camera_ray_over_ground.z >= 0.0
-            || camera_ray_over_ground.x.is_nan()
-            || camera_ray_over_ground.y.is_nan()
-            || camera_ray_over_ground.z.is_nan()
-        {
-            return Err(Error::AboveHorizon);
+        let camera_to_elevated_ground = Isometry3::translation(0., 0., -z) * self.camera_to_ground;
+
+        let camera_position = camera_to_elevated_ground * Point3::origin();
+        let camera_ray_over_ground = camera_to_elevated_ground * camera_ray;
+
+        if relative_eq!(camera_ray_over_ground.z, 0.0) {
+            return Err(Error::NotOnProjectionPlane);
         }
 
-        let distance_to_plane = z - self.camera_to_ground.translation.z;
-        let slope = distance_to_plane / camera_ray_over_ground.z;
-        let intersection_point =
-            self.camera_to_ground.translation.vector + camera_ray_over_ground * slope;
-        Ok(point![intersection_point.x, intersection_point.y])
+        let intersection_scalar = -camera_position.z / camera_ray_over_ground.z;
+
+        if intersection_scalar < 0.0 {
+            return Err(Error::NotOnProjectionPlane);
```

Comment:
I'm not whether this should return `NotOnProjectionPlane` or `BehindCamera`...

Reply:
You're right, `BehindCamera` makes more sense here.